File: crates/stateset-embedded/src/analytics.rs
//! Analytics and forecasting operations

use stateset_core::{
    AnalyticsQuery, CustomerMetrics, DemandForecast, FulfillmentMetrics, InventoryHealth,
    InventoryMovement, LowStockItem, OrderStatusBreakdown, ProductPerformance, Result,
    ReturnMetrics, RevenueByPeriod, RevenueForecast, SalesSummary, TimeGranularity,
    TopCustomer, TopProduct,
};
use stateset_db::Database;
use std::sync::Arc;
use rust_decimal::Decimal;

/// Analytics operations interface.
///
/// Provides sales analytics, inventory forecasting, and business intelligence.
///
/// # Example
///
/// ```rust,no_run
/// use stateset_embedded::{Commerce, AnalyticsQuery, TimePeriod};
///
/// let commerce = Commerce::new("./store.db")?;
///
/// // Get sales summary for last 30 days
/// let summary = commerce.analytics().sales_summary(
///     AnalyticsQuery::new().period(TimePeriod::Last30Days)
/// )?;
/// println!("Revenue: ${}", summary.total_revenue);
///
/// // Get top selling products
/// let top_products = commerce.analytics().top_products(
///     AnalyticsQuery::new().period(TimePeriod::ThisMonth).limit(10)
/// )?;
///
/// // Get demand forecast
/// let forecasts = commerce.analytics().demand_forecast(None, 30)?;
/// # Ok::<(), stateset_embedded::CommerceError>(())
/// ```
pub struct Analytics {
    db: Arc<dyn Database>,
}

impl Analytics {
    pub(crate) fn new(db: Arc<dyn Database>) -> Self {
        Self { db }
    }

    // ========================================================================
    // Sales Analytics
    // ========================================================================

    /// Get sales summary for a time period.
    ///
    /// Returns total revenue, order count, average order value, and more.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # let commerce = Commerce::new(":memory:")?;
    /// let summary = commerce.analytics().sales_summary(
    ///     AnalyticsQuery::new().period(TimePeriod::Last30Days)
    /// )?;
    /// println!("Revenue: ${}", summary.total_revenue);
    /// println!("Orders: {}", summary.order_count);
    /// println!("AOV: ${}", summary.average_order_value);
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn sales_summary(&self, query: AnalyticsQuery) -> Result<SalesSummary> {
        self.db.analytics().get_sales_summary(query)
    }

    /// Get revenue broken down by time periods.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # let commerce = Commerce::new(":memory:")?;
    /// let revenue = commerce.analytics().revenue_by_period(
    ///     AnalyticsQuery::new()
    ///         .period(TimePeriod::Last30Days)
    ///         .granularity(TimeGranularity::Day)
    /// )?;
    /// for day in revenue {
    ///     println!("{}: ${}", day.period, day.revenue);
    /// }
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn revenue_by_period(&self, query: AnalyticsQuery) -> Result<Vec<RevenueByPeriod>> {
        self.db.analytics().get_revenue_by_period(query)
    }

    // ========================================================================
    // Product Analytics
    // ========================================================================

    /// Get top selling products.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # let commerce = Commerce::new(":memory:")?;
    /// let top = commerce.analytics().top_products(
    ///     AnalyticsQuery::new()
    ///         .period(TimePeriod::ThisMonth)
    ///         .limit(10)
    /// )?;
    /// for product in top {
    ///     println!("{}: {} units, ${}", product.name, product.units_sold, product.revenue);
    /// }
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn top_products(&self, query: AnalyticsQuery) -> Result<Vec<TopProduct>> {
        self.db.analytics().get_top_products(query)
    }

    /// Get product performance with period comparison.
    pub fn product_performance(&self, query: AnalyticsQuery) -> Result<Vec<ProductPerformance>> {
        self.db.analytics().get_product_performance(query)
    }

    // ========================================================================
    // Customer Analytics
    // ========================================================================

    /// Get customer metrics.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # let commerce = Commerce::new(":memory:")?;
    /// let metrics = commerce.analytics().customer_metrics(
    ///     AnalyticsQuery::new().period(TimePeriod::ThisMonth)
    /// )?;
    /// println!("Total customers: {}", metrics.total_customers);
    /// println!("New this month: {}", metrics.new_customers);
    /// println!("Avg LTV: ${}", metrics.average_lifetime_value);
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn customer_metrics(&self, query: AnalyticsQuery) -> Result<CustomerMetrics> {
        self.db.analytics().get_customer_metrics(query)
    }

    /// Get top customers by spend.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # let commerce = Commerce::new(":memory:")?;
    /// let top = commerce.analytics().top_customers(
    ///     AnalyticsQuery::new().period(TimePeriod::AllTime).limit(10)
    /// )?;
    /// for customer in top {
    ///     println!("{}: {} orders, ${}", customer.name, customer.order_count, customer.total_spent);
    /// }
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn top_customers(&self, query: AnalyticsQuery) -> Result<Vec<TopCustomer>> {
        self.db.analytics().get_top_customers(query)
    }

    // ========================================================================
    // Inventory Analytics
    // ========================================================================

    /// Get inventory health summary.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # let commerce = Commerce::new(":memory:")?;
    /// let health = commerce.analytics().inventory_health()?;
    /// println!("Total SKUs: {}", health.total_skus);
    /// println!("Low stock: {}", health.low_stock_skus);
    /// println!("Out of stock: {}", health.out_of_stock_skus);
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn inventory_health(&self) -> Result<InventoryHealth> {
        self.db.analytics().get_inventory_health()
    }

    /// Get low stock items.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # use rust_decimal_macros::dec;
    /// # let commerce = Commerce::new(":memory:")?;
    /// let low_stock = commerce.analytics().low_stock_items(Some(dec!(20)))?;
    /// for item in low_stock {
    ///     println!("{}: {} available", item.sku, item.available);
    /// }
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn low_stock_items(&self, threshold: Option<Decimal>) -> Result<Vec<LowStockItem>> {
        self.db.analytics().get_low_stock_items(threshold)
    }

    /// Get inventory movement summary.
    pub fn inventory_movement(&self, query: AnalyticsQuery) -> Result<Vec<InventoryMovement>> {
        self.db.analytics().get_inventory_movement(query)
    }

    // ========================================================================
    // Order Analytics
    // ========================================================================

    /// Get order status breakdown.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # let commerce = Commerce::new(":memory:")?;
    /// let breakdown = commerce.analytics().order_status_breakdown(
    ///     AnalyticsQuery::new().period(TimePeriod::Last30Days)
    /// )?;
    /// println!("Pending: {}", breakdown.pending);
    /// println!("Shipped: {}", breakdown.shipped);
    /// println!("Delivered: {}", breakdown.delivered);
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn order_status_breakdown(&self, query: AnalyticsQuery) -> Result<OrderStatusBreakdown> {
        self.db.analytics().get_order_status_breakdown(query)
    }

    /// Get fulfillment metrics.
    pub fn fulfillment_metrics(&self, query: AnalyticsQuery) -> Result<FulfillmentMetrics> {
        self.db.analytics().get_fulfillment_metrics(query)
    }

    // ========================================================================
    // Return Analytics
    // ========================================================================

    /// Get return metrics.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # let commerce = Commerce::new(":memory:")?;
    /// let metrics = commerce.analytics().return_metrics(
    ///     AnalyticsQuery::new().period(TimePeriod::ThisMonth)
    /// )?;
    /// println!("Returns: {}", metrics.total_returns);
    /// println!("Return rate: {}%", metrics.return_rate_percent);
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn return_metrics(&self, query: AnalyticsQuery) -> Result<ReturnMetrics> {
        self.db.analytics().get_return_metrics(query)
    }

    // ========================================================================
    // Forecasting
    // ========================================================================

    /// Get demand forecast for inventory items.
    ///
    /// Predicts future demand based on historical sales data.
    ///
    /// # Arguments
    ///
    /// * `skus` - Optional list of SKUs to forecast. If None, forecasts all items.
    /// * `days_ahead` - Number of days to forecast ahead.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # let commerce = Commerce::new(":memory:")?;
    /// // Forecast for all items
    /// let forecasts = commerce.analytics().demand_forecast(None, 30)?;
    /// for f in forecasts {
    ///     println!("{}: {} units/day, {} days until stockout",
    ///         f.sku,
    ///         f.average_daily_demand,
    ///         f.days_until_stockout.unwrap_or(999)
    ///     );
    /// }
    ///
    /// // Forecast for specific SKUs
    /// let forecasts = commerce.analytics().demand_forecast(
    ///     Some(vec!["SKU-001".to_string(), "SKU-002".to_string()]),
    ///     14
    /// )?;
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn demand_forecast(
        &self,
        skus: Option<Vec<String>>,
        days_ahead: u32,
    ) -> Result<Vec<DemandForecast>> {
        self.db.analytics().get_demand_forecast(skus, days_ahead)
    }

    /// Get revenue forecast.
    ///
    /// Predicts future revenue based on historical trends.
    ///
    /// # Arguments
    ///
    /// * `periods_ahead` - Number of periods to forecast.
    /// * `granularity` - Time granularity (Day, Week, Month).
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use stateset_embedded::*;
    /// # let commerce = Commerce::new(":memory:")?;
    /// let forecasts = commerce.analytics().revenue_forecast(3, TimeGranularity::Month)?;
    /// for f in forecasts {
    ///     println!("{}: ${} (${} - ${})",
    ///         f.period,
    ///         f.forecasted_revenue,
    ///         f.lower_bound,
    ///         f.upper_bound
    ///     );
    /// }
    /// # Ok::<(), CommerceError>(())
    /// ```
    pub fn revenue_forecast(
        &self,
        periods_ahead: u32,
        granularity: TimeGranularity,
    ) -> Result<Vec<RevenueForecast>> {
        self.db.analytics().get_revenue_forecast(periods_ahead, granularity)
    }
}

File: cli/src/commands/analytics.js
/**
 * Analytics Commands Module
 *
 * Read-oriented analytics and forecasting commands for stateset-direct.
 */

const DEFAULT_PERIOD = 'last30days';

export async function execute(action, args, { commerce, output, jsonOutput }) {
  switch (action) {
    case 'sales': {
      const period = args[0] || DEFAULT_PERIOD;
      const summary = await commerce.analytics.salesSummary({ period });
      return formatSalesSummary(period, summary, { jsonOutput });
    }

    case 'revenue': {
      const period = args[0] || DEFAULT_PERIOD;
      const granularity = args[1] || 'day';
      const rows = await commerce.analytics.revenueByPeriod({ period, granularity });
      return formatRevenueRows(period, granularity, rows, { output, jsonOutput });
    }

    case 'products': {
      const period = args[0] || DEFAULT_PERIOD;
      const limit = Number.parseInt(args[1] || '10', 10);
      const products = await commerce.analytics.topProducts({ period, limit });
      return formatTopProducts(period, products, { output, jsonOutput });
    }

    case 'customers': {
      const period = args[0] || DEFAULT_PERIOD;
      const metrics = await commerce.analytics.customerMetrics({ period });
      return formatCustomerMetrics(period, metrics, { jsonOutput });
    }

    case 'health': {
      const health = await commerce.analytics.inventoryHealth();
      return formatInventoryHealth(health, { jsonOutput });
    }

    case 'low': {
      const threshold = Number.parseInt(args[0] || '10', 10);
      const items = await commerce.analytics.lowStockItems(threshold);
      return formatLowStock(threshold, items, { output, jsonOutput });
    }

    case 'forecast': {
      const daysAhead = Number.parseInt(args[0] || '30', 10);
      const skus = args.slice(1);
      const forecasts = await commerce.analytics.demandForecast(
        skus.length > 0 ? skus : undefined,
        daysAhead,
      );
      return formatForecasts(daysAhead, forecasts, { output, jsonOutput });
    }

    case 'fulfillment': {
      const period = args[0] || DEFAULT_PERIOD;
      const metrics = await commerce.analytics.fulfillmentMetrics({ period });
      return formatFulfillmentMetrics(period, metrics, { jsonOutput });
    }

    case 'returns': {
      const period = args[0] || DEFAULT_PERIOD;
      const metrics = await commerce.analytics.returnMetrics({ period });
      return formatReturnMetrics(period, metrics, { jsonOutput });
    }

    default:
      throw new Error(
        `Unknown action: analytics ${action}\n\n` +
          'Available actions:\n' +
          '  sales [period]               Sales summary\n' +
          '  revenue [period] [granularity]  Revenue over time\n' +
          '  products [period] [limit]    Top products\n' +
          '  customers [period]           Customer metrics\n' +
          '  health                       Inventory health\n' +
          '  low [threshold]              Low stock items\n' +
          '  forecast [days] [sku...]     Demand forecast\n' +
          '  fulfillment [period]         Fulfillment metrics\n' +
          '  returns [period]             Return metrics',
      );
  }
}

function formatSalesSummary(period, summary, { jsonOutput }) {
  if (jsonOutput) return { period, summary };
  return {
    summary,
    formatted:
      `Sales summary (${period})\n` +
      `${'-'.repeat(32)}\n` +
      `Revenue:        ${summary.totalRevenue}\n` +
      `Orders:         ${summary.orderCount}\n` +
      `AOV:            ${summary.averageOrderValue}\n` +
      `Items sold:     ${summary.itemsSold}\n` +
      `Customers:      ${summary.uniqueCustomers}`,
  };
}

function formatRevenueRows(period, granularity, rows, { output, jsonOutput }) {
  if (jsonOutput) return { period, granularity, rows };
  if (rows.length === 0) return { formatted: `No revenue data found for ${period}.` };
  const formatted = output.table(rows, [
    { key: 'period', header: 'Period' },
    { key: 'revenue', header: 'Revenue', align: 'right' },
    { key: 'orders', header: 'Orders', align: 'right' },
  ]);
  return { rows, formatted: `Revenue by ${granularity} (${period})\n\n${formatted}` };
}

function formatTopProducts(period, products, { output, jsonOutput }) {
  if (jsonOutput) return { period, products };
  if (products.length === 0) return { formatted: `No product analytics found for ${period}.` };
  const formatted = output.table(products, [
    { key: 'sku', header: 'SKU' },
    { key: 'name', header: 'Name' },
    { key: 'unitsSold', header: 'Units', align: 'right' },
    { key: 'revenue', header: 'Revenue', align: 'right' },
    { key: 'orderCount', header: 'Orders', align: 'right' },
  ]);
  return { products, formatted: `Top products (${period})\n\n${formatted}` };
}

function formatCustomerMetrics(period, metrics, { jsonOutput }) {
  if (jsonOutput) return { period, metrics };
  return {
    metrics,
    formatted:
      `Customer metrics (${period})\n` +
      `${'-'.repeat(32)}\n` +
      `Total customers:      ${metrics.totalCustomers}\n` +
      `New customers:        ${metrics.newCustomers}\n` +
      `Returning customers:  ${metrics.returningCustomers}\n` +
      `Avg lifetime value:   ${metrics.averageLifetimeValue}\n` +
      `Avg orders/customer:  ${metrics.averageOrdersPerCustomer}`,
  };
}

function formatInventoryHealth(health, { jsonOutput }) {
  if (jsonOutput) return { health };
  return {
    health,
    formatted:
      'Inventory health\n' +
      `${'-'.repeat(24)}\n` +
      `Total SKUs:       ${health.totalSkus}\n` +
      `In stock:         ${health.inStockSkus}\n` +
      `Low stock:        ${health.lowStockSkus}\n` +
      `Out of stock:     ${health.outOfStockSkus}\n` +
      `Inventory value:  ${health.totalValue}`,
  };
}

function formatLowStock(threshold, items, { output, jsonOutput }) {
  if (jsonOutput) return { threshold, items };
  if (items.length === 0) return { formatted: `No low stock items at threshold ${threshold}.` };
  const formatted = output.table(items, [
    { key: 'sku', header: 'SKU' },
    { key: 'name', header: 'Name' },
    { key: 'available', header: 'Available', align: 'right' },
    { key: 'reorderPoint', header: 'Reorder', align: 'right' },
    { key: 'daysOfStock', header: 'Days Left', align: 'right' },
  ]);
  return { items, formatted: `Low stock items (threshold ${threshold})\n\n${formatted}` };
}

function formatForecasts(daysAhead, forecasts, { output, jsonOutput }) {
  if (jsonOutput) return { daysAhead, forecasts };
  if (forecasts.length === 0)
    return { formatted: `No forecasts available for next ${daysAhead} days.` };
  const formatted = output.table(forecasts, [
    { key: 'sku', header: 'SKU' },
    { key: 'forecastDemand', header: 'Forecast', align: 'right' },
    { key: 'currentStock', header: 'Stock', align: 'right' },
    { key: 'daysUntilStockout', header: 'Days to Stockout', align: 'right' },
  ]);
  return { forecasts, formatted: `Demand forecast (${daysAhead} days)\n\n${formatted}` };
}

function formatFulfillmentMetrics(period, metrics, { jsonOutput }) {
  if (jsonOutput) return { period, metrics };
  return {
    metrics,
    formatted:
      `Fulfillment metrics (${period})\n` +
      `${'-'.repeat(35)}\n` +
      `Orders fulfilled:      ${metrics.ordersFulfilled}\n` +
      `Avg fulfill time:      ${metrics.averageFulfillmentTime}\n` +
      `On-time rate:          ${metrics.onTimeRate}\n` +
      `Exception rate:        ${metrics.exceptionRate}`,
  };
}

function formatReturnMetrics(period, metrics, { jsonOutput }) {
  if (jsonOutput) return { period, metrics };
  return {
    metrics,
    formatted:
      `Return metrics (${period})\n` +
      `${'-'.repeat(30)}\n` +
      `Return count:       ${metrics.returnCount}\n` +
      `Return rate:        ${metrics.returnRate}\n` +
      `Refund amount:      ${metrics.totalRefundAmount}\n` +
      `Top reason:         ${metrics.topReason || 'N/A'}`,
  };
}

export const metadata = {
  name: 'analytics',
  aliases: ['an', 'stats'],
  description: 'Analytics and forecasting commands',
  actions: {
    sales: { description: 'Sales summary', args: ['[period]'] },
    revenue: { description: 'Revenue by time period', args: ['[period]', '[granularity]'] },
    products: { description: 'Top products', args: ['[period]', '[limit]'] },
    customers: { description: 'Customer metrics', args: ['[period]'] },
    health: { description: 'Inventory health summary', args: [] },
    low: { description: 'Low stock items', args: ['[threshold]'] },
    forecast: { description: 'Demand forecast', args: ['[days]', '[sku...]'] },
    fulfillment: { description: 'Fulfillment metrics', args: ['[period]'] },
    returns: { description: 'Return metrics', args: ['[period]'] },
  },
};

export default { execute, metadata };
